### PR TITLE
RavenExceptions , EventHandlers, HttpCache

### PR DESCRIPTION
**Pull request by sashatim125**

### CODE REVIEW

File: Raven.CppClient.Tests/ClientConfigurationTest.cpp [hunk 1]
```diff
@@ -16,6 +18,29 @@ namespace ravendb::client::tests::client::documents::operations::configuration
 		}
 	};
 
+	TEST_F(ClientConfigurationTest, CanSaveAndReadServerWideClientConfiguration)
+	{
+		//TODO test is OK , yet it changes global definitions -> waiting for embedded
```
Comment (ayende) [inline Raven.CppClient.Tests/ClientConfigurationTest.cpp:23]: I don't like this.
If you need to skip it, use that, see: https://stackoverflow.com/questions/7208070/googletest-how-to-skip-a-test

File: Raven.CppClient/HttpCache.cpp [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+			item->last_sever_update = std::chrono::steady_clock::now();
+		}
+	}
+
+	std::chrono::milliseconds HttpCache::ReleaseCacheItem::get_age() const
+	{
+		if(!item)
+		{
+#ifdef max
+#undef max
```
Comment (ayende) [inline Raven.CppClient/HttpCache.cpp:24]: why is this?

File: Raven.CppClient/HttpCache.cpp [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+	{
+		auto lock = std::shared_lock(_global_mutex);
+		return _items.size();
+	}
+
+	void HttpCache::set(std::string url, std::string change_vector, std::string result)
+	{
+		if (_total_size += result.size() * sizeof(std::string::value_type) > _max_size)
+		{
+			auto clean = std::async(std::launch::async, [http_cache = _weak_this.lock()]()->void
```
Comment (ayende) [inline Raven.CppClient/HttpCache.cpp:114]: how is this running?

File: Raven.CppClient.Tests/ClientConfigurationTest.cpp (hunk 1 above)
Comment (sashatim125) [reply Raven.CppClient.Tests/ClientConfigurationTest.cpp:23]: OK, done

File: Raven.CppClient/HttpCache.cpp (hunk 2 above)
Comment (sashatim125) [reply Raven.CppClient/HttpCache.cpp:24]: there is a Windows macro `max` -> tried to #undefine it.

File: Raven.CppClient/HttpCache.cpp (hunk 3 above)
Comment (sashatim125) [reply Raven.CppClient/HttpCache.cpp:114]: a new thread is created for this labda execution.

> If the async flag is set (i.e. (policy & std::launch::async) != 0), then async executes the callable object f on a new thread of execution (with all thread-locals initialized) as if spawned by std::thread(std::forward<F>(f), std::forward<Args>(args)...), except that if the function f returns a value or throws an exception, it is stored in the shared state accessible through the std::future that async returns to the caller.

from [https://en.cppreference.com/w/cpp/thread/async](https://en.cppreference.com/w/cpp/thread/async)

**Pull request merged**